tools/tsp-client/src/network.ts

```diff
@@ -0,0 +1,92 @@
+import { createDefaultHttpClient, createPipelineRequest } from "@azure/core-rest-pipeline";
+import { createFileTree } from "./fileTree.js";
+import { resolveImports } from "./typespec.js";
+import { Logger } from "./log.js";
+
+const httpClient = createDefaultHttpClient();
+
+async function fetch(url: string, method: "GET" | "HEAD" = "GET"): Promise<string> {
+  const result = await httpClient.sendRequest(createPipelineRequest({ url, method }));
+  if (result.status !== 200) {
+    throw new Error(`failed to fetch ${url}: ${result.status}`);
+  }
+  return String(result.bodyAsText);
+}
+
+export function isValidUrl(url: string) {
+  try {
+    new URL(url);
+    return true;
+  } catch {
+    return false;
+  }
+}
+
+export function doesFileExist(url: string): Promise<boolean> {
+  return fetch(url, "HEAD")
+    .then(() => true)
+    .catch(() => false);
+}
+
+export function rewriteGitHubUrl(url: string): string {
+  if (url.includes("github.com")) {
+    const result = url.replace("github.com", "raw.githubusercontent.com");
+    Logger.debug(`rewriting github url to direct link: ${result}`);
+    return result;
+  }
+  return url;
+}
+
+export async function downloadTsp(rootUrl: string) {
+  const seenFiles = new Set<string>();
+  const moduleImports = new Set<string>();
+  // fetch the root file
+  const filesToProcess = [rootUrl];
+  seenFiles.add(rootUrl);
+  const fileTree = createFileTree(rootUrl);
+
+  while (filesToProcess.length > 0) {
+    const url = filesToProcess.shift()!;
+    const sourceFile = await fetch(url);
+    fileTree.addFile(url, sourceFile);
```

Comment:
Will this handle relative paths outside of the main directory? That is an expected scenario for any shared tsp projects.

Reply:
It will handle whatever files are relatively referenced so long as they are not on a different URL host. We could extend it to support multiple hosts, but I didn't think it would be likely to be needed for now.